// @apify/consts/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ACTOR_BASE_DOCKER_IMAGES: () => ACTOR_BASE_DOCKER_IMAGES,
  ACTOR_BASE_DOCKER_IMAGE_DEFAULT: () => ACTOR_BASE_DOCKER_IMAGE_DEFAULT,
  ACTOR_CATEGORIES: () => ACTOR_CATEGORIES,
  ACTOR_EVENT_NAMES: () => ACTOR_EVENT_NAMES,
  ACTOR_LIMITS: () => ACTOR_LIMITS,
  ACTOR_LOG_MAX_CHARS: () => ACTOR_LOG_MAX_CHARS,
  ACTOR_NAME: () => ACTOR_NAME,
  ACTOR_RESTART_ON_ERROR: () => ACTOR_RESTART_ON_ERROR,
  ACT_JOB_STATUSES: () => ACT_JOB_STATUSES,
  ACT_JOB_TERMINAL_STATUSES: () => ACT_JOB_TERMINAL_STATUSES,
  ACT_JOB_TYPES: () => ACT_JOB_TYPES,
  ACT_RESTART_ON_ERROR: () => ACT_RESTART_ON_ERROR,
  ACT_SOURCE_TYPES: () => ACT_SOURCE_TYPES,
  ACT_TYPES: () => ACT_TYPES,
  ALL_ACTOR_CATEGORIES: () => ALL_ACTOR_CATEGORIES,
  ANONYMOUS_USERNAME: () => ANONYMOUS_USERNAME,
  APIFY_PROXY_VALUE_REGEX: () => APIFY_PROXY_VALUE_REGEX,
  BUILD_TAG_LATEST: () => BUILD_TAG_LATEST,
  COMMA_SEPARATED_EMAILS_REGEX: () => COMMA_SEPARATED_EMAILS_REGEX,
  COMMA_SEPARATED_EMAILS_REGEX_STR: () => COMMA_SEPARATED_EMAILS_REGEX_STR,
  COMPUTE_UNIT_MB: () => COMPUTE_UNIT_MB,
  COMPUTE_UNIT_MILLIS: () => COMPUTE_UNIT_MILLIS,
  CONTACT_LINK_REGEX: () => CONTACT_LINK_REGEX,
  DEFAULT_CONTAINER_PORT: () => DEFAULT_CONTAINER_PORT,
  DEFAULT_PLATFORM_LIMITS: () => DEFAULT_PLATFORM_LIMITS,
  DNS_SAFE_NAME_MAX_LENGTH: () => DNS_SAFE_NAME_MAX_LENGTH,
  DNS_SAFE_NAME_REGEX: () => DNS_SAFE_NAME_REGEX,
  DOCKER_LABELS: () => DOCKER_LABELS,
  EMAIL_REGEX: () => EMAIL_REGEX,
  EMAIL_REGEX_STR: () => EMAIL_REGEX_STR,
  ENV_VARS: () => ENV_VARS,
  FINISHED_PROJECT_STATUSES: () => FINISHED_PROJECT_STATUSES,
  FREE_SUBSCRIPTION_PLAN_CODE: () => FREE_SUBSCRIPTION_PLAN_CODE,
  GITHUB_GIST_URL_REGEX: () => GITHUB_GIST_URL_REGEX,
  GITHUB_REGEX: () => GITHUB_REGEX,
  GIT_MAIN_BRANCH: () => GIT_MAIN_BRANCH,
  GIT_REPO_REGEX: () => GIT_REPO_REGEX,
  HTTP_URL_REGEX: () => HTTP_URL_REGEX,
  INTEGER_ENV_VARS: () => INTEGER_ENV_VARS,
  KEY_VALUE_STORE_KEYS: () => KEY_VALUE_STORE_KEYS,
  KEY_VALUE_STORE_KEY_REGEX: () => KEY_VALUE_STORE_KEY_REGEX,
  LEGACY_ACTOR_CATEGORIES: () => LEGACY_ACTOR_CATEGORIES,
  LOCAL_ENV_VARS: () => LOCAL_ENV_VARS,
  LOCAL_STORAGE_SUBDIRS: () => LOCAL_STORAGE_SUBDIRS,
  MARKETPLACE_USER_ROLES: () => MARKETPLACE_USER_ROLES,
  MAX_MULTIFILE_BYTES: () => MAX_MULTIFILE_BYTES,
  MAX_PAYLOAD_SIZE_BYTES: () => MAX_PAYLOAD_SIZE_BYTES,
  META_ORIGINS: () => META_ORIGINS,
  ME_USER_NAME_PLACEHOLDER: () => ME_USER_NAME_PLACEHOLDER,
  PROJECT_STATUSES: () => PROJECT_STATUSES,
  PROXY_URL_REGEX: () => PROXY_URL_REGEX,
  RELATIVE_URL_REGEX: () => RELATIVE_URL_REGEX,
  REQUEST_QUEUE_HEAD_MAX_LIMIT: () => REQUEST_QUEUE_HEAD_MAX_LIMIT,
  SHORT_CRAWLER_ID_LENGTH: () => SHORT_CRAWLER_ID_LENGTH,
  SOURCE_FILE_FORMATS: () => SOURCE_FILE_FORMATS,
  SPLIT_PATH_REGEX: () => SPLIT_PATH_REGEX,
  TWITTER_REGEX: () => TWITTER_REGEX,
  URL_REGEX: () => URL_REGEX,
  USERNAME: () => USERNAME,
  USER_BASIC_TEXT_XSS_OPTIONS: () => USER_BASIC_TEXT_XSS_OPTIONS,
  VERSION_INT_MAJOR_BASE: () => VERSION_INT_MAJOR_BASE,
  VERSION_INT_MINOR_BASE: () => VERSION_INT_MINOR_BASE,
  WEBHOOK_ALLOWED_PAYLOAD_VARIABLES: () => WEBHOOK_ALLOWED_PAYLOAD_VARIABLES,
  WEBHOOK_DEFAULT_PAYLOAD_TEMPLATE: () => WEBHOOK_DEFAULT_PAYLOAD_TEMPLATE,
  WEBHOOK_DISPATCH_STATUSES: () => WEBHOOK_DISPATCH_STATUSES,
  WEBHOOK_EVENT_TYPES: () => WEBHOOK_EVENT_TYPES,
  WEBHOOK_EVENT_TYPE_GROUPS: () => WEBHOOK_EVENT_TYPE_GROUPS,
  WORKER_SERVICE_TYPES: () => WORKER_SERVICE_TYPES
});
module.exports = __toCommonJS(src_exports);

// src/regexs.ts
var namePartSubRegexStr = "[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+";
var nameSubRegexStr = `${namePartSubRegexStr}(?:\\.${namePartSubRegexStr})*`;
var domainPartSubRegexStr = "[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?";
var domainSubRegexStr = `${domainPartSubRegexStr}(?:\\.${domainPartSubRegexStr})+`;
var EMAIL_REGEX_STR = `${nameSubRegexStr}@${domainSubRegexStr}`;
var EMAIL_REGEX = new RegExp(`^${EMAIL_REGEX_STR}$`);
var COMMA_SEPARATED_EMAILS_REGEX_STR = `(${EMAIL_REGEX_STR})( *, *${EMAIL_REGEX_STR})*`;
var COMMA_SEPARATED_EMAILS_REGEX = new RegExp(`^${COMMA_SEPARATED_EMAILS_REGEX_STR}$`);
var GIT_REPO_REGEX = /^(?:git|ssh|https?|git@[-\w.]+):(\/\/)?(.*?)(\/?|#[-\d\w._:/]+?)$/;
var DNS_SAFE_NAME_REGEX = /^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])$/;
var APIFY_PROXY_VALUE_REGEX = /^[\w._~]+$/;
var PROXY_URL_REGEX = /^http:\/\/(([^:]+:)?[^@]*@)?[^.:@]+\.[^:]+:[\d]+?$/;
var KEY_VALUE_STORE_KEY_REGEX = /^([a-zA-Z0-9!\-_.'()]{1,256})$/;
var GITHUB_REGEX_STR = "[a-z\\d](?:[a-z\\d]|-(?=[a-z\\d])){0,38}";
var TWITTER_REGEX = /^@[a-z0-9_]{1,15}$/i;
var GITHUB_REGEX = new RegExp(`^${GITHUB_REGEX_STR}$`, "i");
var URL_REGEX = /^https?:\/\//i;
var HTTP_URL_REGEX = new RegExp("^(?:(?:(?:https?):)?\\/\\/)(?:\\S+(?::\\S*)?@)?(?:(?!(?:10|127)(?:\\.\\d{1,3}){3})(?!(?:169\\.254|192\\.168)(?:\\.\\d{1,3}){2})(?!172\\.(?:1[6-9]|2\\d|3[0-1])(?:\\.\\d{1,3}){2})(?:[1-9]\\d?|1\\d\\d|2[01]\\d|22[0-3])(?:\\.(?:1?\\d{1,2}|2[0-4]\\d|25[0-5])){2}(?:\\.(?:[1-9]\\d?|1\\d\\d|2[0-4]\\d|25[0-4]))|(?:(?:[a-z0-9\\u00a1-\\uffff][a-z0-9\\u00a1-\\uffff_-]{0,62})?[a-z0-9\\u00a1-\\uffff]\\.)+(?:[a-z\\u00a1-\\uffff]{2,}\\.?|xn--[a-z0-9]+))(?::\\d{2,5})?(?:[/?#]\\S*)?$", "i");
var GITHUB_GIST_URL_REGEX = new RegExp(`^https:\\/\\/gist\\.github\\.com\\/${GITHUB_REGEX_STR}\\/[0-9a-f]{32}$`, "i");
var SPLIT_PATH_REGEX = /[^/]+/g;
var RELATIVE_URL_REGEX = /^(?!www\.|(?:http|ftp)s?:\/\/|[A-Za-z]:\|\/\/).*/i;
var CONTACT_LINK_REGEX = /^(mailto|tel|sms):.*$/i;

// src/consts.ts
var FREE_SUBSCRIPTION_PLAN_CODE = "DEV";
var ACT_JOB_TYPES = {
  BUILD: "BUILD",
  RUN: "RUN"
};
var ACT_SOURCE_TYPES = {
  SOURCE_CODE: "SOURCE_CODE",
  SOURCE_FILES: "SOURCE_FILES",
  GIT_REPO: "GIT_REPO",
  TARBALL: "TARBALL",
  GITHUB_GIST: "GITHUB_GIST"
};
var ACTOR_EVENT_NAMES = {
  CPU_INFO: "cpuInfo",
  SYSTEM_INFO: "systemInfo",
  MIGRATING: "migrating",
  PERSIST_STATE: "persistState",
  ABORTING: "aborting"
};
var ACT_JOB_STATUSES = {
  READY: "READY",
  RUNNING: "RUNNING",
  SUCCEEDED: "SUCCEEDED",
  FAILED: "FAILED",
  TIMING_OUT: "TIMING-OUT",
  TIMED_OUT: "TIMED-OUT",
  ABORTING: "ABORTING",
  ABORTED: "ABORTED"
};
var WEBHOOK_DISPATCH_STATUSES = {
  ACTIVE: "ACTIVE",
  SUCCEEDED: "SUCCEEDED",
  FAILED: "FAILED"
};
var ACT_JOB_TERMINAL_STATUSES = [
  ACT_JOB_STATUSES.SUCCEEDED,
  ACT_JOB_STATUSES.FAILED,
  ACT_JOB_STATUSES.TIMED_OUT,
  ACT_JOB_STATUSES.ABORTED
];
var WORKER_SERVICE_TYPES = {
  CRAWLING: "crawling",
  ACTOR: "actor"
};
var META_ORIGINS = {
  DEVELOPMENT: "DEVELOPMENT",
  WEB: "WEB",
  API: "API",
  SCHEDULER: "SCHEDULER",
  TEST: "TEST"
};
var ACTOR_BASE_DOCKER_IMAGES = [
  {
    name: "apify/actor-node",
    displayName: "Node.js 14 on Alpine Linux",
    prePull: true
  },
  {
    name: "apify/actor-node-puppeteer-chrome",
    displayName: "Node.js 14 + Puppeteer + Chrome on Debian",
    copyChown: "myuser:myuser",
    prePull: true
  },
  {
    name: "apify/actor-node-playwright-chrome",
    displayName: "Node.js 14 + Playwright + Chrome on Debian",
    copyChown: "myuser:myuser",
    prePull: true
  },
  {
    name: "apify/actor-node-playwright-firefox",
    displayName: "Node.js 14 + Playwright + Firefox on Debian",
    copyChown: "myuser:myuser",
    prePull: true
  },
  {
    name: "apify/actor-node-playwright-webkit",
    displayName: "Node.js 14 + Playwright + WebKit on Debian",
    copyChown: "myuser:myuser",
    prePull: true
  },
  {
    name: "apify/actor-node-playwright",
    displayName: "Node.js 14 + Playwright + All Browsers on Ubuntu",
    copyChown: "myuser:myuser",
    prePull: true
  },
  {
    name: "apify/actor-node",
    displayName: "BETA: Node.js 14 on Alpine Linux:beta"
  },
  {
    name: "apify/actor-node-puppeteer-chrome:beta",
    displayName: "BETA: Node.js 14 + Puppeteer + Chrome on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-playwright-chrome:beta",
    displayName: "BETA: Node.js 14 + Playwright + Chrome on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-playwright-firefox:beta",
    displayName: "BETA: Node.js 14 + Playwright + Firefox on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-playwright-webkit:beta",
    displayName: "BETA: Node.js 14 + Playwright + WebKit on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-playwright:beta",
    displayName: "Node.js 14 + Playwright + All Browsers on Ubuntu",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-basic",
    displayName: "[DEPRECATED]: Node.js 12 on Alpine Linux"
  },
  {
    name: "apify/actor-node-chrome",
    displayName: "[DEPRECATED]: Node.js 12 + Chrome on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-chrome-xvfb",
    displayName: "[DEPRECATED]: Node.js 12 + Chrome + Xvfb on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-basic:beta",
    displayName: "[DEPRECATED] BETA: Node.js 12 on Alpine Linux"
  },
  {
    name: "apify/actor-node-chrome:beta",
    displayName: "[DEPRECATED] BETA: Node.js 12 + Chrome on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-chrome-xvfb:beta",
    displayName: "[DEPRECATED] BETA: Node.js 12 + Chrome + Xvfb on Debian",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-basic:v0.21.10",
    displayName: "[DEPRECATED]: Node.js 12 on Alpine Linux (Apify SDK v0.21.10)"
  },
  {
    name: "apify/actor-node-chrome:v0.21.10",
    displayName: "[DEPRECATED]: Node.js 12 + Chrome on Debian (Apify SDK v0.21.10)",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-chrome-xvfb:v0.21.10",
    displayName: "[DEPRECATED]: Node.js 12 + Chrome + Xvfb on Debian (Apify SDK v0.21.10)",
    copyChown: "myuser:myuser"
  },
  {
    name: "apify/actor-node-puppeteer",
    displayName: "[DEPRECATED] Node.js 10 + Puppeteer on Debian - use apify/actor-node-chrome instead!",
    copyChown: "node:node"
  },
  {
    name: "apify/actor-node-puppeteer:beta",
    displayName: "[DEPRECATED] BETA: Node.js 10 + Puppeteer on Debian - use apify/actor-node-chrome:beta instead!",
    copyChown: "node:node"
  }
];
var ACTOR_BASE_DOCKER_IMAGE_DEFAULT = ACTOR_BASE_DOCKER_IMAGES[0].name;
var DOCKER_LABELS = {
  ACTOR_BUILD_ID: "com.apify.actBuildId",
  ACTOR_RUN_ID: "com.apify.actRunId",
  ACT_BUILD_ID: "com.apify.actBuildId",
  ACT_RUN_ID: "com.apify.actRunId"
};
var ACT_TYPES = {
  ACT: "acts",
  CRAWLER: "crawlers"
};
var ANONYMOUS_USERNAME = "anonymous";
var USERNAME = {
  MIN_LENGTH: 3,
  MAX_LENGTH: 30,
  REGEX: /^[a-zA-Z0-9_.-]{3,30}$/
};
var DNS_SAFE_NAME_MAX_LENGTH = 63;
var ACTOR_NAME = {
  MIN_LENGTH: 3,
  MAX_LENGTH: DNS_SAFE_NAME_MAX_LENGTH,
  REGEX: DNS_SAFE_NAME_REGEX
};
var SHORT_CRAWLER_ID_LENGTH = 5;
var BUILD_TAG_LATEST = "latest";
var ACTOR_RESTART_ON_ERROR = {
  MAX_RESTARTS: 3,
  INTERVAL_MILLIS: 1 * 60 * 1e3
};
var ACT_RESTART_ON_ERROR = ACTOR_RESTART_ON_ERROR;
var COMPUTE_UNIT_MB = 1024;
var COMPUTE_UNIT_MILLIS = 60 * 60 * 1e3;
var ACTOR_LIMITS = {
  BUILD_DEFAULT_MEMORY_MBYTES: 2048,
  BUILD_TIMEOUT_SECS: 600,
  RUN_DISK_TO_MEMORY_SIZE_COEFF: 2,
  RUN_MEMORY_MBYTES_PER_CPU_CORE: 4096,
  FREE_ACCOUNT_MAX_MEMORY_MBYTES: 8192,
  PAID_ACCOUNT_MAX_MEMORY_MBYTES: 65536,
  MIN_RUN_MEMORY_MBYTES: 128,
  MAX_RUN_MEMORY_MBYTES: 32768,
  INPUT_SCHEMA_MAX_BYTES: 100 * 1024,
  LOG_MAX_CHARS: 5e6
};
var DEFAULT_PLATFORM_LIMITS = {
  MAX_ACTORS_PER_USER: 100,
  MAX_TASKS_PER_USER: 1e3,
  MAX_SCHEDULES_PER_USER: 100,
  MAX_WEBHOOKS_PER_USER: 100,
  FREE_ACCOUNT_MAX_CONCURRENT_ACTOR_RUNS_PER_USER: 25,
  PAID_ACCOUNT_MAX_CONCURRENT_ACTOR_RUNS_PER_USER: 250,
  MAX_ACTORS_PER_SCHEDULER: 10,
  MAX_TASKS_PER_SCHEDULER: 10
};
var ME_USER_NAME_PLACEHOLDER = "me";
var REQUEST_QUEUE_HEAD_MAX_LIMIT = 1e3;
var ENV_VARS = {
  IS_AT_HOME: "APIFY_IS_AT_HOME",
  ACTOR_ID: "APIFY_ACTOR_ID",
  ACTOR_RUN_ID: "APIFY_ACTOR_RUN_ID",
  ACTOR_TASK_ID: "APIFY_ACTOR_TASK_ID",
  ACTOR_BUILD_ID: "APIFY_ACTOR_BUILD_ID",
  ACTOR_BUILD_NUMBER: "APIFY_ACTOR_BUILD_NUMBER",
  INPUT_KEY: "APIFY_INPUT_KEY",
  USER_ID: "APIFY_USER_ID",
  TOKEN: "APIFY_TOKEN",
  PROXY_PASSWORD: "APIFY_PROXY_PASSWORD",
  PROXY_HOSTNAME: "APIFY_PROXY_HOSTNAME",
  PROXY_PORT: "APIFY_PROXY_PORT",
  PROXY_STATUS_URL: "APIFY_PROXY_STATUS_URL",
  STARTED_AT: "APIFY_STARTED_AT",
  TIMEOUT_AT: "APIFY_TIMEOUT_AT",
  DEFAULT_KEY_VALUE_STORE_ID: "APIFY_DEFAULT_KEY_VALUE_STORE_ID",
  DEFAULT_DATASET_ID: "APIFY_DEFAULT_DATASET_ID",
  DEFAULT_REQUEST_QUEUE_ID: "APIFY_DEFAULT_REQUEST_QUEUE_ID",
  LOCAL_STORAGE_DIR: "APIFY_LOCAL_STORAGE_DIR",
  API_BASE_URL: "APIFY_API_BASE_URL",
  API_PUBLIC_BASE_URL: "APIFY_API_PUBLIC_BASE_URL",
  HEADLESS: "APIFY_HEADLESS",
  XVFB: "APIFY_XVFB",
  MEMORY_MBYTES: "APIFY_MEMORY_MBYTES",
  LOG_LEVEL: "APIFY_LOG_LEVEL",
  ACTOR_EVENTS_WS_URL: "APIFY_ACTOR_EVENTS_WS_URL",
  CHROME_EXECUTABLE_PATH: "APIFY_CHROME_EXECUTABLE_PATH",
  CONTAINER_PORT: "APIFY_CONTAINER_PORT",
  CONTAINER_URL: "APIFY_CONTAINER_URL",
  META_ORIGIN: "APIFY_META_ORIGIN",
  FACT: "APIFY_FACT",
  DEDICATED_CPUS: "APIFY_DEDICATED_CPUS",
  SDK_LATEST_VERSION: "APIFY_SDK_LATEST_VERSION",
  DISABLE_OUTDATED_WARNING: "APIFY_DISABLE_OUTDATED_WARNING",
  WORKFLOW_KEY: "APIFY_WORKFLOW_KEY",
  ACT_ID: "APIFY_ACT_ID",
  ACT_RUN_ID: "APIFY_ACT_RUN_ID"
};
var INTEGER_ENV_VARS = [
  ENV_VARS.PROXY_PORT,
  ENV_VARS.MEMORY_MBYTES,
  ENV_VARS.CONTAINER_PORT
];
var DEFAULT_CONTAINER_PORT = 4321;
var LOCAL_STORAGE_SUBDIRS = {
  datasets: "datasets",
  keyValueStores: "key_value_stores",
  requestQueues: "request_queues"
};
var LOCAL_ENV_VARS = {
  [ENV_VARS.DEFAULT_KEY_VALUE_STORE_ID]: "default",
  [ENV_VARS.DEFAULT_DATASET_ID]: "default",
  [ENV_VARS.DEFAULT_REQUEST_QUEUE_ID]: "default",
  [ENV_VARS.PROXY_HOSTNAME]: "proxy.apify.com",
  [ENV_VARS.PROXY_PORT]: 8e3 .toString(),
  [ENV_VARS.CONTAINER_PORT]: DEFAULT_CONTAINER_PORT.toString(),
  [ENV_VARS.CONTAINER_URL]: `http://localhost:${DEFAULT_CONTAINER_PORT}`
};
var KEY_VALUE_STORE_KEYS = {
  INPUT: "INPUT",
  OUTPUT: "OUTPUT"
};
var ACTOR_LOG_MAX_CHARS = ACTOR_LIMITS.LOG_MAX_CHARS;
var MAX_PAYLOAD_SIZE_BYTES = 9437184;
var ACTOR_CATEGORIES = {
  AUTOMATION: "Automation",
  BUSINESS: "Business",
  COVID_19: "Covid-19",
  DEVELOPER_EXAMPLES: "Developer examples",
  DEVELOPER_TOOLS: "Developer tools",
  ECOMMERCE: "E-commerce",
  GAMES: "Games",
  JOBS: "Jobs",
  MARKETING: "Marketing",
  NEWS: "News",
  SEO_TOOLS: "SEO tools",
  SOCIAL_MEDIA: "Social media",
  TRAVEL: "Travel",
  VIDEOS: "Videos",
  REAL_ESTATE: "Real estate",
  SPORTS: "Sports",
  EDUCATION: "Education",
  OTHER: "Other"
};
var LEGACY_ACTOR_CATEGORIES = {
  TRAVEL: "Travel",
  ECOMMERCE: "E-commerce",
  ENTERTAINMENT: "Entertainment",
  SOCIAL: "Social",
  MARKETING: "Marketing",
  NEWS: "News",
  FINANCE: "Finance",
  LIFESTYLE: "Lifestyle",
  SEARCH_ENGINES: "Search engines",
  DATA: "Data processing",
  EGOVERNMENT: "E-government",
  TOOLS: "Tools",
  EXAMPLES: "Examples",
  OTHER: "Other"
};
var ALL_ACTOR_CATEGORIES = {
  ...ACTOR_CATEGORIES,
  ...LEGACY_ACTOR_CATEGORIES
};
var VERSION_INT_MAJOR_BASE = 1e7;
var VERSION_INT_MINOR_BASE = 1e5;
var USER_BASIC_TEXT_XSS_OPTIONS = {
  whiteList: {
    a: ["href", "title", "target"],
    code: [],
    strong: [],
    b: [],
    br: [],
    ul: [],
    li: [],
    ol: [],
    i: [],
    u: [],
    p: []
  }
};
var WEBHOOK_EVENT_TYPES = {
  ACTOR_RUN_CREATED: "ACTOR.RUN.CREATED",
  ACTOR_RUN_SUCCEEDED: "ACTOR.RUN.SUCCEEDED",
  ACTOR_RUN_FAILED: "ACTOR.RUN.FAILED",
  ACTOR_RUN_TIMED_OUT: "ACTOR.RUN.TIMED_OUT",
  ACTOR_RUN_ABORTED: "ACTOR.RUN.ABORTED",
  ACTOR_RUN_RESURRECTED: "ACTOR.RUN.RESURRECTED",
  ACTOR_BUILD_CREATED: "ACTOR.BUILD.CREATED",
  ACTOR_BUILD_SUCCEEDED: "ACTOR.BUILD.SUCCEEDED",
  ACTOR_BUILD_FAILED: "ACTOR.BUILD.FAILED",
  ACTOR_BUILD_TIMED_OUT: "ACTOR.BUILD.TIMED_OUT",
  ACTOR_BUILD_ABORTED: "ACTOR.BUILD.ABORTED",
  TEST: "TEST"
};
var WEBHOOK_EVENT_TYPE_GROUPS = {
  ACTOR_RUN: [
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_CREATED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_SUCCEEDED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_FAILED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_TIMED_OUT,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_ABORTED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_RESURRECTED
  ],
  ACTOR_BUILD: [
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_CREATED,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_SUCCEEDED,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_FAILED,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_TIMED_OUT,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_ABORTED
  ],
  ACTOR_RUN_TERMINAL: [
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_SUCCEEDED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_FAILED,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_TIMED_OUT,
    WEBHOOK_EVENT_TYPES.ACTOR_RUN_ABORTED
  ],
  ACTOR_BUILD_TERMINAL: [
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_SUCCEEDED,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_FAILED,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_TIMED_OUT,
    WEBHOOK_EVENT_TYPES.ACTOR_BUILD_ABORTED
  ]
};
var WEBHOOK_DEFAULT_PAYLOAD_TEMPLATE = `{
    "userId": {{userId}},
    "createdAt": {{createdAt}},
    "eventType": {{eventType}},
    "eventData": {{eventData}},
    "resource": {{resource}}
}`;
var WEBHOOK_ALLOWED_PAYLOAD_VARIABLES = /* @__PURE__ */ new Set([
  "userId",
  "createdAt",
  "eventType",
  "eventData",
  "resource"
]);
var MAX_MULTIFILE_BYTES = 3 * 1024 ** 2;
var SOURCE_FILE_FORMATS = {
  TEXT: "TEXT",
  BASE64: "BASE64"
};
var PROJECT_STATUSES = {
  REQUEST: "REQUEST",
  SPECIFICATION: "SPECIFICATION",
  OFFERS: "OFFERS",
  DEPOSIT: "DEPOSIT",
  DEPOSIT_PAID: "DEPOSIT_PAID",
  NEW: "NEW",
  IN_PROGRESS: "IN_PROGRESS",
  QA: "QA",
  CUSTOMER_QA: "CUSTOMER_QA",
  READY_FOR_INVOICE: "READY_FOR_INVOICE",
  INVOICED: "INVOICED",
  PAID: "PAID",
  DELIVERED: "DELIVERED",
  CLOSED: "CLOSED",
  FINISHED: "FINISHED"
};
var FINISHED_PROJECT_STATUSES = [
  PROJECT_STATUSES.READY_FOR_INVOICE,
  PROJECT_STATUSES.INVOICED,
  PROJECT_STATUSES.PAID,
  PROJECT_STATUSES.DELIVERED,
  PROJECT_STATUSES.FINISHED
];
var MARKETPLACE_USER_ROLES = {
  DEVELOPER: "DEVELOPER",
  DATA_EXPERT: "DATA_EXPERT",
  CUSTOMER: "CUSTOMER"
};
var GIT_MAIN_BRANCH = "main";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACTOR_BASE_DOCKER_IMAGES,
  ACTOR_BASE_DOCKER_IMAGE_DEFAULT,
  ACTOR_CATEGORIES,
  ACTOR_EVENT_NAMES,
  ACTOR_LIMITS,
  ACTOR_LOG_MAX_CHARS,
  ACTOR_NAME,
  ACTOR_RESTART_ON_ERROR,
  ACT_JOB_STATUSES,
  ACT_JOB_TERMINAL_STATUSES,
  ACT_JOB_TYPES,
  ACT_RESTART_ON_ERROR,
  ACT_SOURCE_TYPES,
  ACT_TYPES,
  ALL_ACTOR_CATEGORIES,
  ANONYMOUS_USERNAME,
  APIFY_PROXY_VALUE_REGEX,
  BUILD_TAG_LATEST,
  COMMA_SEPARATED_EMAILS_REGEX,
  COMMA_SEPARATED_EMAILS_REGEX_STR,
  COMPUTE_UNIT_MB,
  COMPUTE_UNIT_MILLIS,
  CONTACT_LINK_REGEX,
  DEFAULT_CONTAINER_PORT,
  DEFAULT_PLATFORM_LIMITS,
  DNS_SAFE_NAME_MAX_LENGTH,
  DNS_SAFE_NAME_REGEX,
  DOCKER_LABELS,
  EMAIL_REGEX,
  EMAIL_REGEX_STR,
  ENV_VARS,
  FINISHED_PROJECT_STATUSES,
  FREE_SUBSCRIPTION_PLAN_CODE,
  GITHUB_GIST_URL_REGEX,
  GITHUB_REGEX,
  GIT_MAIN_BRANCH,
  GIT_REPO_REGEX,
  HTTP_URL_REGEX,
  INTEGER_ENV_VARS,
  KEY_VALUE_STORE_KEYS,
  KEY_VALUE_STORE_KEY_REGEX,
  LEGACY_ACTOR_CATEGORIES,
  LOCAL_ENV_VARS,
  LOCAL_STORAGE_SUBDIRS,
  MARKETPLACE_USER_ROLES,
  MAX_MULTIFILE_BYTES,
  MAX_PAYLOAD_SIZE_BYTES,
  META_ORIGINS,
  ME_USER_NAME_PLACEHOLDER,
  PROJECT_STATUSES,
  PROXY_URL_REGEX,
  RELATIVE_URL_REGEX,
  REQUEST_QUEUE_HEAD_MAX_LIMIT,
  SHORT_CRAWLER_ID_LENGTH,
  SOURCE_FILE_FORMATS,
  SPLIT_PATH_REGEX,
  TWITTER_REGEX,
  URL_REGEX,
  USERNAME,
  USER_BASIC_TEXT_XSS_OPTIONS,
  VERSION_INT_MAJOR_BASE,
  VERSION_INT_MINOR_BASE,
  WEBHOOK_ALLOWED_PAYLOAD_VARIABLES,
  WEBHOOK_DEFAULT_PAYLOAD_TEMPLATE,
  WEBHOOK_DISPATCH_STATUSES,
  WEBHOOK_EVENT_TYPES,
  WEBHOOK_EVENT_TYPE_GROUPS,
  WORKER_SERVICE_TYPES
});
